internal/controller/pkg/revision/runtime_function.go

```diff
@@ -0,0 +1,178 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package revision
+
+import (
+	"context"
+	"fmt"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	pkgmetav1beta1 "github.com/crossplane/crossplane/apis/pkg/meta/v1beta1"
+	v1 "github.com/crossplane/crossplane/apis/pkg/v1"
+	"github.com/crossplane/crossplane/apis/pkg/v1beta1"
+	"github.com/crossplane/crossplane/internal/initializer"
+	"github.com/crossplane/crossplane/internal/xpkg"
+)
+
+const (
+	errNotFunction                            = "not a function package"
+	errDeleteFunctionDeployment               = "cannot delete function package deployment"
+	errDeleteFunctionSA                       = "cannot delete function package service account"
+	errApplyFunctionDeployment                = "cannot apply function package deployment"
+	errApplyFunctionSecret                    = "cannot apply function package secret"
+	errApplyFunctionSA                        = "cannot apply function package service account"
+	errApplyFunctionService                   = "cannot apply function package service"
+	errFmtUnavailableFunctionDeployment       = "function package deployment is unavailable with message: %s"
+	errNoAvailableConditionFunctionDeployment = "function package deployment has no condition of type \"Available\" yet"
+)
+
+// FunctionHooks performs runtime operations for function packages.
+type FunctionHooks struct {
+	client resource.ClientApplicator
+}
+
+// NewFunctionHooks returns a new FunctionHooks.
+func NewFunctionHooks(client client.Client) *FunctionHooks {
+	return &FunctionHooks{
+		client: resource.ClientApplicator{
+			Client:     client,
+			Applicator: resource.NewAPIPatchingApplicator(client),
+		},
+	}
+}
+
+// Pre performs operations meant to happen before establishing objects.
+func (h *FunctionHooks) Pre(ctx context.Context, _ runtime.Object, pr v1.PackageRevisionWithRuntime, manifests ManifestBuilder) error {
+	// TODO(ezgidemirel): update any status fields relevant to package revisions.
+
+	if pr.GetDesiredState() != v1.PackageRevisionActive {
+		return nil
+	}
+
+	// Ensure Prerequisites
+	svc := manifests.Service(functionServiceOverrides()...)
```

Comment:
```suggestion
	svc := build.Service(functionServiceOverrides()...)
```

Just a naming suggestion. Or we could keep `manifests` and change the method to be `manifests.BuildService`.